src/Compiler/Microsoft.AspNetCore.Razor.Language/src/RazorDiagnostic.cs

```diff
@@ -1,63 +1,93 @@
 ﻿// Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
-#nullable disable
-
 using System;
+using Microsoft.Extensions.Internal;
 
 namespace Microsoft.AspNetCore.Razor.Language;
 
-public abstract class RazorDiagnostic : IEquatable<RazorDiagnostic>, IFormattable
+public sealed class RazorDiagnostic : IEquatable<RazorDiagnostic>, IFormattable
 {
-    internal static readonly RazorDiagnostic[] EmptyArray = Array.Empty<RazorDiagnostic>();
-    internal static readonly object[] EmptyArgs = Array.Empty<object>();
+    // Internal for testing
+    internal RazorDiagnosticDescriptor Descriptor { get; }
+
+    // Internal for testing
+    internal object[] Args { get; }
 
-    public abstract string Id { get; }
+    public string Id => Descriptor.Id;
+    public RazorDiagnosticSeverity Severity => Descriptor.Severity;
+    public SourceSpan Span { get; }
 
-    public abstract RazorDiagnosticSeverity Severity { get; }
+    private RazorDiagnostic(RazorDiagnosticDescriptor descriptor, SourceSpan? span, object[]? args)
+    {
+        Descriptor = descriptor ?? throw new ArgumentNullException(nameof(descriptor));
+        Span = span ?? SourceSpan.Undefined;
+        Args = args ?? Array.Empty<object>();
+    }
 
-    public abstract SourceSpan Span { get; }
+    public static RazorDiagnostic Create(RazorDiagnosticDescriptor descriptor)
+        => new(descriptor, span: null, args: null);
 
-    public abstract string GetMessage(IFormatProvider formatProvider);
+    public static RazorDiagnostic Create(RazorDiagnosticDescriptor descriptor, SourceSpan? span)
+        => new(descriptor, span, args: null);
 
-    public string GetMessage() => GetMessage(null);
+    public static RazorDiagnostic Create(RazorDiagnosticDescriptor descriptor, params object[] args)
+        => new(descriptor, span: null, args);
 
-    public abstract bool Equals(RazorDiagnostic other);
+    public static RazorDiagnostic Create(RazorDiagnosticDescriptor descriptor, SourceSpan? span, params object[] args)
+        => new(descriptor, span, args);
 
-    public abstract override int GetHashCode();
+    public string GetMessage()
+        => GetMessage(formatProvider: null);
 
-    public static RazorDiagnostic Create(RazorDiagnosticDescriptor descriptor, SourceSpan span)
+    public string GetMessage(IFormatProvider? formatProvider)
+        => string.Format(formatProvider, Descriptor.MessageFormat, Args);
+
+    public override bool Equals(object? obj)
+        => obj is RazorDiagnostic diagnostic
+            ? Equals(diagnostic)
+            : false;
+
+    public bool Equals(RazorDiagnostic other)
```

Comment:
Looks like `IEquatable<T>.Equals` should handle when `other` is `null`: https://learn.microsoft.com/en-us/dotnet/api/system.iequatable-1.equals?view=net-7.0#system-iequatable-1-equals(-0)

Reply:
Good catch! This is a spot where netstandard2.0 didn't help me. :smile: